Report per-MethodInstance inference timings during a single inference call, via SnoopCompile

NHDaly opened the pull request:
This is a second attempt, after https://github.com/JuliaLang/julia/pull/37535 was proving far too complicated.

This PR adds a small bit of code to Core.Compiler that provides Timers, which can be used to build nested timers, much like TimerOutputs.jl, but rewritten from scratch to get it to work inside Bootstrap.

This then optionally enables these timers inside `Core.Compiler.typeinf()` whenever a global const boolean Ref, `__measure_typeinf__`,  is set to `true`.

The timers construct a nested trace of the time to run type inference for each individual invocation of `typeinf()`. These timers are returned as a tree structure, where each node contains the *exclusive* time for that invocation.

For example:
```julia
julia> using SnoopCompileCore

julia> module A
           foo() = 1
           bar() = foo() + 1
       end
Main.A

julia> timing = SnoopCompileCore.@snoopi_deep begin
           A.bar()
       end
Core.Compiler.Timings.Timing("root", 0x00037e9353825b46, 0x00037e9353879a8b, 0x000000000000be6c, Core.Compiler.Timings.Timing[Core.Compiler.Timings.Timing(Tuple{typeof(Main.A.bar)}, 0x00037e9353831cc8, 0x00037e9353864e03, 0x000000000002274a, Core.Compiler.Timings.Timing[Core.Compiler.Timings.Timing(Tuple{typeof(Main.A.foo)}, 0x00037e9353837468, 0x00037e9353837468, 0x0000000000013457, Core.Compiler.Timings.Timing[]), Core.Compiler.Timings.Timing(Tuple{typeof(+),Int64,Int64}, 0x00037e93538541c8, 0x00037e93538541c8, 0x0000000000010b44, Core.Compiler.Timings.Timing[])])])
```

Great care is taken in this implementation to record _only_ the exclusive time for each invocation, in order to not include any of the overhead of the measurement itself in the timings for any individual node. This is important since we don't want to disproportionately report overly large exclusive times for method instances that simply fan out to many children, since this isn't actually reflective of how long it took to infer _that method itself_. We ultimately want to be able to produce _both_ accurate inclusive times _and_ accurate exclusive times, so we record the exclusive times, and take care to not include the overhead, and then reconstruct the inclusive times in post-processing (this code is in SnoopCompile.jl). This algorithm is basically:
 - First, record start time of "root"
 - `typeinf`:
   - `enter`:
     - record stop time of parent node
     - push new node for new invocation
     - record start time of new node
   - invoke `_typeinf`  # May recurse into `typeinf()`
   - `exit`:
     - record stop time of new node
     - pop new node
     - record (continuation) start time of parent node
- At very end, record stop time of "root"

Each Timing node contains its payload information about the invocation (currently just the MethodInstance -- see comment below where I'm asking for suggestions on what else to include), the cumulative exclusive duration for that node, the absolute start time for that node (for ProfileView), and the children from that node. 




I was trying to follow the same mechanism as is done in `@snoopi` - to allow swapping out the `typeinf()` function to a timed version, but it turns out that the overhead of calling `invokelatest` is probably too high.

NHDaly committed:
First pass at providing full inferrable set to SnoopCompile `@snoopi`

NHDaly committed:
Use Array{Bool,0} to control snoopi "collect inference callees" feature

Since Ref{} isn't available yet from inside Core.Compiler :)

NHDaly committed:
Start recording dependency graph edges for snoopi

- I _think_ this should get us all of the MethodInstances that are being
  inferred during inference of a caller function.
- The old approach was missing some edges.
- Also, this now presents that info via a dependency graph, which can be
  used to reverse-engineer which methods are responsible for what time

NHDaly committed:
Comment out old attempt

NHDaly committed:
Second attempt at per-method inference timings, this time by also allowing substituting `typeinf()`

NHDaly committed:
Undo commented out attempted ccall ptr for `typeinf()`

NHDaly committed:
Delete leftovers from previous attempt

NHDaly committed:
Use a 0-dim array for `const _typeinf_function`.

This complicates the code a bit, but I think it should reduce overhead
slightly.

NHDaly committed:
Switch from `_apply_latest()` to `_apply_pure()` which works even inside generated functions

NHDaly committed:
simplify array construction

timholy commented:
Seems like another option would be some `const are_we_snooping = Ref(false)` and modify the inference code to check it, and if true then push the MethodInstance onto your list.

NHDaly commented:
Thanks @timholy  - i had exactly the same thought! That's what i've just finished trying now, and i think it's working. i'll push that up here so you can see what i did.

I wasn't sure if a `Ref{Bool}` or a `Array{Bool,0}` is preferable - do you know what the recommended way to have a global, mutable, const bool is in Core.Compiler?

NHDaly committed:
Implement very basic nested timing in Core.Compiler.Timings

This allows us to measure the _nested_ per-method-instance inference
timings, which allows us to produce both a profile view over the
inference data, as well as compute _exclusive_ timings
per-method-instance.

NHDaly commented:
The problem, then, of course, is that I had to implement a very basic version of TimerOutputs in the Core.Compiler bootstrap, which can keep track of the nested timings, but I think i've done that. it wasn't too hard!

NHDaly committed:
Add start_time to Core.Compiler.Timings to support profiles

NHDaly committed:
Simplify Timing() constructor

NHDaly committed:
Exclude snoop overhead from recursive per-method inference timings

Manually stop and restart the timers before and after calling child
functions, so that we directly compute _exclusive_ timings for each node
in the call tree.

NHDaly committed:
Fix root timing: reset() sets current time instead of 0

NHDaly committed:
Performance-tweaking typeinf measurement overhead.

`@inbounds` and rewriting to be more efficient.

Moved the timings as close to the `invoke _typeinf` as possible, to
reduce measurement of overhead.  But it didn't work: the measured timing
still covers the whole process.

NHDaly committed:
Extract out enter_new_timer(name), exit_current_timer(_expected_name_) to simplify code.

NHDaly committed:
Inline typeinf timing helper functions. Simplify __measure_typeinf__ bool creation

NHDaly commented:
Okay! :) I believe this analysis is working! We've been using it internally at RelationalAI, and it's pretty cool, and has already proven informative. 😁 

We'd love to get this merged this week. I'll try to focus on it to make sure it happens!

Here's the remaining work I see:
- [ ] Settle on exactly what information to record for each MethodInstance.
   - The `linfo` / MethodInstance object itself?
     - Good: contains relevant info about which method and which argument types, plus lots of other stuff.
     - Good: cheap to copy (just a pointer).
     - Bad: Cannot be serialized.
     - Bad: Somewhat opaque.
     - Bad: doesn't capture recursive call info / constant propagation info
   - A TypeTuple (`Tuple{typeof(f), Int, Int}`)? 
     - Good: can be serialized.
     - Good: Contains function type and arg types.
     - Bad: loses function instance info (i.e. not correct for closures or other callable objects).
     - Bad: doesn't capture recursive call info / constant propagation info (same as `linfo`)
   - A two-tuple of (f, tt): (`(+, (Int, Int))`).
      - I like this approach, it's best of both the above worlds i think.
   - Maybe something that contains more info about the invocation? Like, when we compile the same MethodInstance multiple times, *why* are we doing that? Likely we're specializing it for different constants? Or there's some recursion stuff involved? This would be good to record.
    - I guess this info is in the InferenceState. What info should we pull out?
- [ ] Performance improvements
   - Figure out best kind of "global const boolean" -- Array{0,Bool} or Ref{Bool}?
   - Try to get the overhead from snooping is as low as possible when snooping is enabled.
- [ ] Clean up the inner "Timings" module 
   - Is there any way this can be done without adding so much code? If not, what should we name that module?

NHDaly committed:
Use at-inboudnds more consistently in Core.Compiler.Timings

NHDaly commented:
After a good amount of profiling, I have convinced myself that we already have the overhead during snooping _quite_ reasonably low.

For example, to snoop `@code_typed peakflops()`, we see only 2 / 3806 samples in this measurement, both within the call to `time_ns()`:
```julia
julia> @profile timing = SnoopCompileCore.@snoopi_deep begin
           @code_typed peakflops()
       end;
```
Then from PProf:
```
  Total:           2       3806 (flat, cum) 99.69%
    126            .          .           function typeinf(interp::AbstractInterpreter, frame::InferenceState) 
    127            .          .               if __measure_typeinf__[] 
    128            .          1                   Timings.enter_new_timer(frame.linfo.specTypes) 
    129            .          .            
    130            .       3799                   v = _typeinf(interp, frame) 
    131            .          .            
    132            1          1                   Timings.exit_current_timer(frame.linfo.specTypes) 
    133            .          .            
    134            1          1                   return v 
    135            .          .               else 
    136            .          4                   return _typeinf(interp, frame) 
    137            .          .               end 
    138            .          .           end 
    139            .          .            
    140            .          .           function _typeinf(interp::AbstractInterpreter, frame::InferenceState) 
    141            .          .               typeinf_nocycle(interp, frame) || return false # frame is now part of a higher cycle 
```
```
Active filters:
   focus=enter_new_timer
Showing nodes accounting for 1, 100% of 1 total
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context 	 	 
----------------------------------------------------------+-------------
                                                 1   100% |   Core.Compiler.typeinf /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:128 (inline)
         0     0%     0%          1   100%                | enter_new_timer /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:43
                                                 1   100% |   time_ns /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:25 (inline)
----------------------------------------------------------+-------------
```
```
Active filters:
   focus=exit_current_timer
Showing nodes accounting for 1, 100% of 1 total
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context 	 	 
----------------------------------------------------------+-------------
                                                 1   100% |   Core.Compiler.typeinf /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:132 (inline)
         0     0%     0%          1   100%                | exit_current_timer /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:84
                                                 1   100% |   time_ns /Users/nathandaly/src/julia/usr/share/julia/base/compiler/typeinfer.jl:25 (inline)
----------------------------------------------------------+-------------
```

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,134 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+
+function close_current_timer()
+    stop_time = time_ns()
+    parent_timer = _timings[end]
```
I guess this works because inference uses a depth-first search?

NHDaly replied:
Yeah, exactly, you're right.

I _believe_ that it's using a depth-first search, since everything seems to be implemented via recursive calls, rather than via an explicit queue. I'm pretty sure that's true. But you're totally right that this wouldn't work if it switched to some other traversal order. That's definitely worth calling out in comments; lemme do that now. :) thanks.

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,134 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
 function typeinf(interp::AbstractInterpreter, frame::InferenceState)
+    if __measure_typeinf__[]
+        Timings.enter_new_timer(frame.linfo.specTypes)
+
```
Maybe reduce the amount of whitespace here?

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,134 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+
 function typeinf(interp::AbstractInterpreter, frame::InferenceState)
+    if __measure_typeinf__[]
+        Timings.enter_new_timer(frame.linfo.specTypes)
```
Personally I would probably use the MethodInstance itself, mostly for the sheer convenience of being able to do `mi.def` to recover the method (which is easier and more performant than calling `which`). But I understand your concerns about the tradeoffs.

NHDaly replied:
I think you're right. We can always extract a serializable format from that in post-processing! :) It seems better to collect the most useful info up front.

Per @vchuravy's suggestion, i've also added the current world age to what's being recorded.

NHDaly replied:
(Done in latest push)

timholy reviewed (commented):
To me this mostly seems good, and that of your list of remaining tasks only the first needs to be resolved; the performance seems unlikely to be an issue (and your measurements prove it), and to me your timing code seems reasonable. Maybe @KristofferC (as author of TimerOutputs) would have more to say, but I'd say that you can safely convert this from draft to "real PR" form and I expect we can merge this quite soon.

NHDaly committed:
Fix close_current_timer() "root" timings. Add comments.

Adds more comments throughout the Core.Compiler.Timings module.

NHDaly committed:
Switch to recording MethodInstance, and world, in Core.Compiler.Timings.

NHDaly commented:
Thanks Tim! :) Okay, then I guess this is mostly complete!

The only thing I'd like to address is maybe also adding some information to `mi_info` that identifies if a function is being specialized for constants, and which ones, since that might help understand a function that's being inferred repeatedly. I think that's in the `slottypes`? I'll add that now. And maybe the `sptypes`? 😬 i'm worried this is just gonna be collecting half the info in the frame!

timholy commented:
Does this help?
```julia
julia> foo(a::AbstractArray{T,N}) where {T,N} = "hi"
foo (generic function with 1 method)

julia> foo([0x0a, 0x0b])
"hi"

julia> using MethodAnalysis

julia> mi = methodinstances(foo)[1]
MethodInstance for foo(::Vector{UInt8})

julia> mi.specTypes
Tuple{typeof(foo), Vector{UInt8}}

julia> mi.sparam_vals
svec(UInt8, 1)

```
Together these already collect the `slottypes` (`specType`) and `sparams` (`sparam_vals`).

NHDaly commented:
:) Thanks for the pointer, @timholy!

No, I looking to capture more than that: Sometimes a method will be specialized for a _constant_, which means we may end up inferring it multiple times. Here's an example that I'm now capturing in this PR after adding the above change:

```julia
julia> using SnoopCompileCore
[ Info: Precompiling SnoopCompileCore [e2b509da-e806-4183-be48-004708413034]

julia> module C
          i(x) = x === 0 ? 0 : 1 / x
          
          a(x) = i(0) * i(x)
          b() = i(0)
          c() = i(0) * i(1) * i(2)
       end;

julia> timing = SnoopCompileCore.@snoopi_deep begin
           C.b()
           C.a(2)
           C.c()
       end;

julia> [t for t in SnoopCompileCore.flatten_times(timing)[1:end-1] if t.second[1].def.name == :i]
6-element Vector{Pair{Float64,Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}}}:
   6.0871e-5 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Core.Const(0, false)])
   7.1293e-5 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Core.Const(0, false)])
   9.5957e-5 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Core.Const(2, false)])
 0.000102533 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Core.Const(1, false)])
 0.000112385 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Core.Const(0, false)])
 0.000208184 => (MethodInstance for i(::Int64), 0x0000000000006f4a, [], [Int64])
```
You can see that `C.i(x)` was inferred 6 different times. One for `::Int`, and then one each for each of the invocations with a constant. But in each case, the `MethodInstance` is the same, so the `MethodInstance` itself doesn't capture this information.

```julia
julia> unique([t.second[1] for t in SnoopCompileCore.flatten_times(timing)[1:end-1] if t.second[1].def.name == :i])
1-element Vector{Core.MethodInstance}:
 MethodInstance for i(::Int64)
```

I _am_ surprised that it repeated the invocations for `i(0)` each time, but maybe it would be too expensive to cache the inference results for every function against every constant, so Julia makes a tradeoff and decides it's better to infer it repeatedly?

^ But I think the above information is useful to know if e.g. you see the same MethodInstance showing up hundreds of times in your inference results, and you want to know why. :)

NHDaly commented:
And for completeness, the const info doesn't show up in the shared method info, of course:
```julia
julia> mi = ans[1]
MethodInstance for i(::Int64)

julia> mi.specTypes
Tuple{typeof(Main.C.i),Int64}

julia> mi.sparam_vals
svec()
```

NHDaly commented:
Yeah, this is really interesting! I didn't know that this was how it worked before. It makes sense, I think (at least, i can produce a rationalization for why it might behave this way), but I didn't really _know_ that we didn't cache specializations for constants. This is useful information to know, I think! :)

```julia
julia> timing = SnoopCompileCore.@snoopi_deep begin
           @eval begin
               f() = 2+2
               f()
           end
       end;

julia> SnoopCompileCore.flatten_times(timing)
3-element Vector{Any}:
 0.000104038 => (MethodInstance for +(::Int64, ::Int64), 0x0000000000006f58, Any[Core.Const(Int64, false)], Any[Core.Const(2, false), Core.Const(2, false)])
 0.000120799 => (MethodInstance for f(), 0x0000000000006f58, Any[], Any[])
 0.000472764 => "root"

julia> timing = SnoopCompileCore.@snoopi_deep begin
           @eval begin
               f() = 2+2
               f()
           end
       end;

julia> SnoopCompileCore.flatten_times(timing)
3-element Vector{Any}:
 0.000101811 => (MethodInstance for f(), 0x0000000000006f59, Any[], Any[])
 0.000109713 => (MethodInstance for +(::Int64, ::Int64), 0x0000000000006f59, Any[Core.Const(Int64, false)], Any[Core.Const(2, false), Core.Const(2, false)])
 0.000467622 => "root"
```

NHDaly committed:
Record more information in typeinf timing:

- MethodInstance
- Current world counter
- Parameter Types being paramaterized on
- Argument Types being inferred on (including constants)

NHDaly commented:
Hmmmmm, this is interesting, and makes me think I might be missing recursive inference cases...

Does inference through cycles not call `typeinf()`?

NHDaly committed:
Add tests of basic typeinf time functionality introduced in #37749

NHDaly committed:
Merge remote-tracking branch 'origin/master' into nhd-snoopi-inferrable-set--TimerOutputs-profile

NHDaly commented:
:) I think this PR is ready for review! :)

Please take a look when you get a chance! Thanks!

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+    return nothing
+end
+
+@inline function exit_current_timer(_expected_frame_)
```
Any reason not to `_expected_frame_` -> `frame`? The name seems to be signaling something but I'm not getting it.

NHDaly replied:
Thanks, good point. I think what I was trying trying to signal was that `frame` is not _actually_ needed at this step of the algorithm, i'm literally only passing it here for the error checking in the `@assert` to make sure that our assumptions hold (the assumption about DFS you noted in your previous review).

I could probably make that more explicit with a comment? Do you have any preferences? :) thanks for reviewing!

timholy replied:
A comment would be fine, but not a big deal either way.

timholy reviewed (commented):
LGTM. If no one else objects I can merge as-is or after any changes to address my small comments.

NHDaly committed:
Rename s/Core.Compiler.__toggle_measure_typeinf/Core.Compiler.__set_measure_typeinf/g

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+    Core.Compiler.__toggle_measure_typeinf(false)
+Toggle recording per-method-instance timings within type inference in the Compiler.
+"""
+__toggle_measure_typeinf(onoff::Bool) = __measure_typeinf__[] = onoff
```
Yes, good suggestion. Lemme change that now.

Done in b61a64ace 👍

NHDaly reviewed (commented):
:) Thanks for the review, @timholy!

NHDaly commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+"""
+function reset_timings()
+    Core.Compiler.empty!(_timings)
+    Core.Compiler.push!(_timings, Timing("root", _time_ns()))
```
Oh! Wait, one thing i'd like to fix before merging is to make the root node have a consistent type as all the other nodes.

Currently the other nodes' `mi_info`s are `Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}`.

I had them as an `::Any` because i was still hacking on it. lemme change it to this, but then i'm not sure what to put for the MethodInstance for `"root"`....

Maybe I could make a dummy method instance for a dummy "root" function or something? Any other suggestion?

timholy replied:
Do you really need a dummy root? `close_current_timer` and `exit_current_timer` could check to see if the stack is empty. Or do they need to record something in some parent somewhere? If you really do need a dummy node, perhaps make it a `Union{Nothing, Tuple{...}}` and just set it to `nothing`?

NHDaly replied:
Hrmmmmm, maybe you're right! Good suggestion! I think it was mainly there to make the algorithm simpler, but you're right we probably don't need it!

It's currently storing the complement of the inference time, which is interesting philosophically, but I think we can reconstruct that in post-processing. Great suggestion, i'll delete this today. :)

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,179 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+"""
+    Core.Compiler.__set_measure_typeinf(true)
+    Core.Compiler.__set_measure_typeinf(false)
+Toggle recording per-method-instance timings within type inference in the Compiler.
```
```suggestion
If set to `true`, record per-method-instance timings within type inference in the Compiler.
```

NHDaly commented:
Thanks again for the review, @timholy! I'll get to your last comment today.

Also, since you're looking at this, would you mind also reviewing https://github.com/JuliaLang/julia/pull/37136/? I want to ensure they're both merged before the 1.6 freeze. :) Jameson had a look 3 weeks ago and gave his SGTM, but it took me a really long time to respond to his comments. 🙈 Thanks so much! :)

NHDaly committed:
Improve comments in base/compiler/typeinfer.jl

NHDaly committed:
Use a consistent type for Timings.mi_info (`Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}`)

NHDaly commented:
> Thanks again for the review, @timholy! I'll get to your last comment today.

Okay, done! :) Thanks

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+
+# What we record for any given frame we infer during type inference.
+function _typeinf_identifier(frame::Core.Compiler.InferenceState)
+    # TODO: What information do we want to collect to identify each invocation of type inference?
```
TODO taken care of? :)

NHDaly replied:
<3 thanks!

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+ROOT()  # Call it to compile a method instance for it.
+"""
+    Core.Compiler.reset_timings()
+Empty out the previously recorded type inference timings, and start the ROOT() timer again.
```
Perhaps add the conventional empty line between the signature and docstring body, and maybe directly identify `Core.Compiler._timings` in the body, e.g. "Empty out the previously recorded type inference timings (`Core.Compiler._timings`), and start..."? :)

NHDaly replied:
Thanks! done.

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+    push!(_timings, Timing(
+        (first(Core.Compiler.methods(ROOT).ms[1].specializations), 0x0, Any[], Any[],),
+        _time_ns()))
+    nothing
```
Perhaps explicit `return nothing`? :)

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+end
+reset_timings()
+
+# (This is split into a function so that it can be called both here and once at the Very
```
Perhaps `here` -> `at the top of enter_new_timer`? :)

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+
+@inline function enter_new_timer(frame)
+    # Very first thing, stop the active timer: get the current time and add in the
+    # time since it was last started to its aggregate exclussive time.
```
Exclus**s**ive :)

Sacha0 commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,189 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+end  # module Timings
+
+"""
+    Core.Compiler.__set_measure_typeinf(true)
```
Perhaps `Core.Compiler.__set_measure_typeinf(onoff::Bool)` followed by an empty line? :)

Sacha0 reviewed (approved):
Fantastic! Thanks @NHDaly! :)

NHDaly committed:
Docstring, comment, and code cleanups from PR review

Co-Authored-By: Sacha Verweij <sacha@stanford.edu>

NHDaly reviewed (commented):
Thank you @Sacha0 for the careful review! :) I've applied your suggested changes. 👍 thanks!

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,192 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+MethodInstance.
+"""
+struct Timing
+    mi_info::Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}
```
Now that this is no longer `Any`, I think it would be better to create a `struct` or `NamedTuple` for this type. The reason: meaning-by-position (i.e., `x[2]` is the method, `x[3]` is the type-tuple, ...) is neither transparent nor future-proof. Aside from avoiding stupid bugs, if we use a `struct` or `NamedTuple` SnoopCompile can also support multiple Julia (future) versions which make different choices, just by using `hasfield(Core.Compiler.Timing.InferenceInstanceInfo, :foo)`. In general I find multiversion support using `hasmethod` and `hasfield` to be more robust than `if Base.VERSION >= v"X.Y.Z"`, particularly while *developing* new features,  since version-checks only work for merged PRs.

I'm not attached to `InferenceInstanceInfo` as a name, use what you think is best.

NHDaly replied:
Thanks good suggestion! I've gone with `InferenceFrameInfo`, which seems to capture everything we mean by it?

Lemme know what you think, and if you have any other suggestions :) Thanks!

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,192 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+# ROOT() is an empty function used as the top-level Timing node to measure all time spent
+# *not* in type inference during a given recording trace. It is used as a "dummy" node.
+function ROOT() end
+ROOT()  # Call it to compile a method instance for it.
```
```suggestion
const ROOTmi = Core.Compiler.specialize_method(first(methods(ROOT)), Tuple{typeof(ROOT)}, Core.svec())
```
and then just use `ROOTmi` directly in `reset_timings`.

timholy commented on base/compiler/typeinfer.jl:
```diff
@@ -8,7 +8,192 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
+reset_timings()
+
+# (This is split into a function so that it can be called both in this module, at the top
+# of `enter_new_timer()``, and once at the Very End of the operation, by whoever started
```
```suggestion
# of `enter_new_timer()`, and once at the Very End of the operation, by whoever started
```

timholy reviewed (commented):
Sorry to ask for more stuff, but I do think this is the last before we merge.

timholy commented:
> would you mind also reviewing #37136, I want to ensure they're both merged before the 1.6 freeze

I strongly approve of the idea of #37136, and it looks pretty straightforward, but I'd be lying if I said I was the right person to review it. I've barely glanced at the LLVM side of compilation. If you've gotten approvals from others I think it should be good to merge.

NHDaly committed:
PR feedback: use a Struct for snoopi info: InferenceFrameInfo

Simplify ROOT MethodInfo construction

Sacha0 reviewed (approved):
Looks great! Thanks Nathan! :)

timholy commented:
Thanks for sticking with it, Nathan! Looking forward to this.

NHDaly commented:
:) Thanks @timholy! I appreciate your help guiding it to the finish! I'm looking forward to this too!

timholy commented:
I've played with this a bit. Could we add a flag so that only *new* inferences are tracked? I piped the output to a file via AbstractTrees and I'm getting lots of stuff like
```
│     └─ MethodInstance for Attributes(::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:color, :strokewidth, :raw), Tuple{Observable{Any}, Int64, Bool}}})
│        └─ MethodInstance for collect(::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:color, :strokewidth, :raw), Tuple{Observable{Any}, Int64, Bool}}})
│           ├─ MethodInstance for (::Colon)(::Int64, ::Int64)
│           │  └─ MethodInstance for UnitRange{Int64}(::Int64, ::Int64)
│           │     ├─ MethodInstance for convert(::Type{Int64}, ::Int64)
│           │     └─ MethodInstance for unitrange_last(::Int64, ::Int64)
│           │        ├─ MethodInstance for >=(::Int64, ::Int64)
│           │        │  └─ MethodInstance for <=(::Int64, ::Int64)
│           │        ├─ MethodInstance for -(::Int64, ::Int64)
│           │        └─ MethodInstance for convert(::Type{Int64}, ::Int64)
```
I count a total of 75 instances of `(::Colon)(::Int64, ::Int64)` just up to depth 10. If folks want to browse a text-dump they will definitely appreciate a shorter list.

timholy commented:
Perhaps like you, I am also a bit puzzled about the results. Why do we have a bazillion
```julia
│  ├─ 0.072064: MethodInstance for (::Colon)(::Int64, ::Int64)
│  │  └─ 0.110615: MethodInstance for UnitRange{Int64}(::Int64, ::Int64)
│  │     ├─ 0.046747: MethodInstance for convert(::Type{Int64}, ::Int64)
│  │     └─ 0.134257: MethodInstance for unitrange_last(::Int64, ::Int64)
│  │        ├─ 0.036451: MethodInstance for -(::Int64, ::Int64)
│  │        └─ 0.019563: MethodInstance for convert(::Type{Int64}, ::Int64)
```
where the times are printed in milliseconds? This suggests it's spending 100μs inferring (presumably looking up from cache) the `UnitRange{Int64}` constructor as called by `(::Colon)(::Int64, ::Int64)`? If you've already inferred the return type of that call, why do you have to repeatedly dive into the callees?

I seem to remember that you opened an issue about this, but I am not finding it.

NHDaly commented:
> Could we add a flag so that only _new_ inferences are tracked?

@timholy I am 60% sure that _indeed_ only new inferences are being tracked.

That is, from what I understand, we are only entering this function if we need to infer something and it's not already in the cache. At least, that's what I intended. Did I not do that right?

My impression is that indeed the results you're seeing are from _actual_ duplicated inference; likely because of inferring constants! I did raise a question about this here, but unfortunately we ended up discussing it offline so the results aren't recorded:
https://github.com/JuliaLang/julia/pull/37535#issuecomment-692450116

My understanding is that what you're seeing can be explained by the following:
- We **do** cache MethodInstances specialized for a constant _within_ a given inferable set. But we **do not** cache MethodInstances specialized for a constant _across_ inferable sets.
- And, we sometimes have to infer the same instance multiple times within recursion, though i do not understand this case well still.

NHDaly commented:
I think we should augment the AbstractTrees view to include the `Const()` info in the printed message for each node; maybe by replacing that argument with the Const?

timholy commented:
See the modified `printnode` method [here](https://julialang.slack.com/archives/C688QKS7Q/p1603127862185200).

NHDaly commented:
Ah yeah, perfect!
